lib/cbdocument.utils.js

```diff
@@ -131,40 +131,82 @@ exports.buildRefValues = function (indexes, data) {
       const path = indexes[v].path;
       const name = indexes[v].name;
       const indexType = indexes[v].indexType;
-      const value = mpath.get(path, data);
-      let valToStore = null;
-      if (typeof value === 'string') {
-        valToStore = value;
-      }
-      if (typeof value === 'number') {
-        valToStore = value.toString();
-      }
-      if (typeof value === 'object' && value._isBaseModel) {
-        valToStore = value.getDocumentKeyValue(value.config.storeFullReferenceId);
-      } else if (Array.isArray(value)) {
-        valToStore = _.map(value, v => {
-          if (v) {
-            if (typeof v === 'string') {
-              return v;
-            }
-            if (typeof v === 'number') {
-              return v.toString();
-            }
-            if (typeof v === 'object' && v._isBaseModel) {
-              return v.getDocumentKeyValue(v.config.storeFullReferenceId);
+
+      const processValue = initialValue => {
```

Comment:
if this doesn't need to be here due to lexical closure or something else (from initial look it doesn't seem to be) I would say just make it into a separate function within utils.